Add SmartBFT consensus support

Param-S opened the pull request:
#### Type of change
- New feature

#### Description
Add support for BFT consensus to orderer consensus plugin

#### Additional details
RFC: https://hyperledger.github.io/fabric-rfcs/text/006-bft-based-ordering-service.html

#### Related issues
EPIC task: #3771 

Signed-off-by: Yacov Manevich <yacovm@il.ibm.com>
Signed-off-by: Yoav Tock [tock@il.ibm.com](mailto:tock@il.ibm.com)
Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
add SMaRtBFT consensus support

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

yacovm commented on go.sum:
```diff
@@ -34,6 +34,8 @@ github.com/Microsoft/hcsshim v0.8.14/go.mod h1:NtVKoYxQuTLx6gEq0L96c9Ju4JbRJ4nY2
 github.com/OneOfOne/xxhash v1.2.2/go.mod h1:HSdplMjZKSmBqAxg5vPj2TmRDmfkzw+cTzAElWljhcU=
 github.com/Shopify/sarama v1.19.0/go.mod h1:FVkBWblsNy7DGZRfXLU0O9RCGt5g3g3yEuWXgklEdEo=
 github.com/Shopify/toxiproxy v2.1.4+incompatible/go.mod h1:OXgGpZ6Cli1/URJOF1DMxUHB2q5Ap20/P/eIdh4G0pI=
+github.com/SmartBFT-Go/consensus v0.3.0 h1:C7PHKU6K6DWpMpK93asWRcX/yqjJ38oYNl57JcJIqsY=
```
what branch is that?

Param-S replied:
It should be the latest. `go get github.com/SmartBFT-Go/consensus@latest`

yacovm replied:
I think that some commits of the default branch (v2) were not included. Can you point it to the latest commit hash instead?

Param-S replied:
Yes. updated

Param-S committed:
Upgraded smartbft consensus library

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
smartbft mock interfaces

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
smartbft consensus unit tests

Signed-off-by: Yacov Manevich yacovm@il.ibm.com
Signed-off-by: Yoav Tock tock@il.ibm.com
Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

yacovm committed:
Validate block validation policy matches consenter set

Change-Id: Ia0f926ddb984cc3008542328a85d41af56be36d3
Signed-off-by: Yacov Manevich <yacovm@il.ibm.com>

Param-S committed:
bft configtxgen updates

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
integration test updates

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
update to orderer main with debug statements

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

yacovm committed:
BFT chains are up and running

Change-Id: I59bf40ff90be2299aa1733cc672ea88f94984b4f
Signed-off-by: Yacov Manevich <yacovm@il.ibm.com>

andrew-coleman committed:
WIP first integration test

DO NOT MERGE - the signature verification in the peer has been temporarily disabled

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

yacovm committed:
Only append block signature if needed

Change-Id: I13ebe069fe4f8fb9b7a3d4671668a5da85cab8ea
Signed-off-by: Yacov Manevich <yacovm@il.ibm.com>

yacovm committed:
Use identifier headers instead of signature headers and properly encode/decode signatures

Change-Id: Ifdd7e9a2a690dc6328bfec60ad22efae9a42fa7c
Signed-off-by: Yacov Manevich <yacovm@il.ibm.com>

Param-S committed:
integration test for update config transaction

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

yacovm committed:
Properly encode BFT policy when updating consenters

Change-Id: I9eb09e2874f4c4fbea77d99aa0528204412c0820
Signed-off-by: Yacov Manevich <yacovm@il.ibm.com>

Param-S committed:
block puller failing while block data identity referencing

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

andrew-coleman committed:
Fix signature verification

Change the BlockPuller to use the protoutil.BlockVerifierFunc generator which works for BFT blocks.

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

Param-S committed:
reconfig with node removal failing

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

andrew-coleman committed:
Add “smartbft multiple nodes view change” test

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

Param-S committed:
integration test iteratively add and remove nodes

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
updated orderer rejoin integration test scenario

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
fixed lint erros and upgrade consensus module

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

andrew-coleman committed:
Fix unit tests

Most of the test build failures were due to changing the signature verifier from an interface to a function generated by a HOF.  This meant that the mocks in the tests had to be rewritten.
There are still some test failures in the smartbft package that I suspect haven’t worked in this version of Fabric.  They will be addressed in another commit.

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

andrew-coleman committed:
Fix ConfigTx ConsenterMapping integration test

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

andrew-coleman committed:
Fix raft integration tests

The Raft channel_participation integration tests were failing because the new logic in initializeMultichannelRegistrar() for initialising without system channel assumes that the consenter type is BFT unless it is told otherwise in the local config.
This commit adds the consenter type to the orderer config template

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

andrew-coleman committed:
Fix linter error

Fixes CI failure in BasicChecks phase

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

andrew-coleman committed:
Fix failures in smartbft unit tests

Many of these were failing due to comparing error message that contain serialized proto structures.  Ideally, these shouldn’t be relied on in tests.

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

andrew-coleman committed:
Add smartbft integration test to CI

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

andrew-coleman committed:
Fix configtx integration test

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

yacovm committed:
Align log expectation in test to actual log printed by the BFT library

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Fix autonomous synchronization integration test

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Make smartbft node addition and removal pass

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm commented on common/policies/bft.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	sp := &cb.SignaturePolicyEnvelope{
+		Rule:       policydsl.NOutOf(int32(2*f+1), pols),
```
I just realized that this is not correct. We need to compute the quorum which is the smallest set size such that every two such sets intersect at f+1 common nodes.

Param-S replied:
Updated it

yacovm commented on integration/nwo/components.go:
```diff
@@ -72,5 +71,5 @@ const CCEnvDefaultImage = "hyperledger/fabric-ccenv:latest"
 var RequiredImages = []string{
 	CCEnvDefaultImage,
-	runner.CouchDBDefaultImage,
+	// runner.CouchDBDefaultImage,
```
please remove

yacovm commented on orderer/common/onboarding/onboarding_test.go:
```diff
@@ -1030,39 +1034,39 @@ func TestVerifierLoader(t *testing.T) {
-				"mychannel": verifier,
-			},
-		},
+		// {
```
please don't leave commented out code

yacovm commented on orderer/common/server/main.go:
```diff
@@ -845,6 +889,48 @@ func initializeEtcdraftConsenter(consenters map[string]consensus.Consenter, conf
+	getConfigBlock := func() *cb.Block {
+		return multichannel.ConfigBlockOrPanic(systemLedger)
+	}
+	icr := onboarding.NewInactiveChainReplicator(ri, getConfigBlock, ri.RegisterChain, conf.General.Cluster.ReplicationBackgroundRefreshInterval)
```
we don't have this inactive chain replication thingy anymore, because we are using channel participation.

Param-S replied:
updated it

yacovm commented on orderer/common/multichannel/ledger_resources_test.go:
```diff
@@ -73,55 +71,55 @@ func (mrm *mutableResourcesMock) CreateBundle(channelID string, c *common.Config
-	// Scenario V: A valid config envelope is passed
-	require.NoError(t, cs.VerifyBlockSignature([]*protoutil.SignedData{}, testConfigEnvelope(t)))
-}
+// func TestLedgerResources_VerifyBlockSignature(t *testing.T) {
```
why is this commented out?

Param-S committed:
code cleanup and updated quorum calculation

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
go modules conflict

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
Removed commneted code

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

yacovm commented on common/crypto/sanitize.go:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	// We assume that the consenter and the CA use the same signature scheme.
+	curveOrderUsedByCryptoGen := cert.PublicKey.(*ecdsa.PublicKey).Curve.Params().N
```
This code will fail if the certificate is not signed by ECDSA. 

Let's proceed but open an issue for this.

Param-S replied:
Opened the issue to track this https://github.com/hyperledger/fabric/issues/4007

yacovm commented on common/crypto/sanitize.go:
```diff
@@ -0,0 +1,106 @@
+		return nil, errors.Wrapf(err, "failed unmarshaling certificate")
+	}
+
+	newSig, err := utils.MarshalECDSASignature(r, s)
```
we have this code in other places in Fabric, we should probably use that code instead of re-writing it. Let's proceed but open an issue.

Param-S replied:
https://github.com/hyperledger/fabric/issues/4007 is updated to track this problem too.

yacovm commented on common/crypto/signer.go:
```diff
@@ -0,0 +1,68 @@
+}
+
+// NewSignatureHeaderCreator creates new signature headers
+func NewSignatureHeaderCreator(ss SignerSupport) *SignatureHeaderCreator {
```
this is un-used.

Param-S replied:
removed it.

yacovm commented on common/policies/bft.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	sp := &cb.SignaturePolicyEnvelope{
+		Rule:       policydsl.NOutOf(int32(math.Ceil(float64(n+f+1)/2)), pols),
```
Let's abstract this quorum computation into a different function, ok?

Param-S replied:
done

yacovm commented on integration/nwo/components.go:
```diff
@@ -72,5 +71,4 @@ const CCEnvDefaultImage = "hyperledger/fabric-ccenv:latest"
 
 var RequiredImages = []string{
 	CCEnvDefaultImage,
-	runner.CouchDBDefaultImage,
```
let's not remove images we don't need.

yacovm commented on integration/nwo/network.go:
```diff
@@ -789,7 +796,7 @@ func (n *Network) Bootstrap() {
 
 	n.bootstrapIdemix()
 
-	if n.SystemChannel != nil { // TODO this entire block could be removed once we finish using the system channel
```
why do we need this change?

yacovm commented on integration/nwo/template/configtx_template.go:
```diff
@@ -158,19 +158,16 @@ Profiles:{{ range .Profiles }}
         PreferredMaxBytes: 512 KB
       {{- end}}
       Capabilities:
-        {{- if eq $w.Consensus.Type "BFT" }}
```
why are you deleting this?

tock-ibm replied:
Why indeed?

yacovm commented on integration/nwo/template/configtx_template.go:
```diff
@@ -231,6 +226,8 @@ Profiles:{{ range .Profiles }}
         Endorsement:
           Type: ImplicitMeta
           Rule: "MAJORITY Endorsement"
+    {{- if .Consortium }}
```
don't we have this change already in the main?

yacovm commented on integration/nwo/template/orderer_template.go:
```diff
@@ -62,6 +62,7 @@ Consensus:
   WALDir: {{ .OrdererDir Orderer }}/etcdraft/wal
   SnapDir: {{ .OrdererDir Orderer }}/etcdraft/snapshot
   EvictionSuspicion: 5s
+  Type: {{ $w.Consensus.Type }}
```
why do we need this?

andrew-coleman replied:
That was added in [this commit ](https://github.com/hyperledger/fabric/pull/3781/commits/f9d8e0d2e83affffab4e29929f2666fa7e10e384)to fix the raft integration tests.

yacovm commented on orderer/common/cluster/util.go:
```diff
@@ -348,42 +336,21 @@ func VerifyBlockHash(indexInBuffer int, blockBuff []*common.Block) error {
-		if err != nil {
-			return nil, errors.Errorf("failed unmarshalling signature header for block with id %d: %v",
-				block.Header.Number, err)
+func searchConsenterIdentityByID(consenters []*common.Consenter, identifier uint32) []byte {
```
This is not used anymore

yacovm commented on orderer/common/cluster/util_test.go:
```diff
@@ -183,71 +183,6 @@ func TestStandardDialer(t *testing.T) {
 	require.ErrorContains(t, err, "error adding root certificate")
 }
 
-func TestVerifyBlockSignature(t *testing.T) {
```
Now this test has been deleted and nothing checks the method any longer.

yacovm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -1149,3 +1168,28 @@ func channelNameFromConfigTx(configtx *cb.Envelope) (string, error) {
+	r.lock.RUnlock()
+
+	if !exists {
+		// This is for the system channel
```
we don't have a system channel any longer

yacovm commented on orderer/consensus/smartbft/configverifier.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+func (cbv *ConfigBlockValidator) checkConsentersMatchPolicy(conf *cb.Config) error {
+	// TODO: check nil pointers everywhere
```
Now would be a good time taking care of the TODO

yacovm commented on orderer/consensus/etcdraft/util.go:
```diff
@@ -465,3 +465,36 @@ func CreateConsentersMap(blockMetadata *etcdraft.BlockMetadata, configMetadata *
 	return consenters
 }
+
+func CreateX509VerifyOptions(ordererConfig channelconfig.Orderer) (x509.VerifyOptions, error) {
```
This is not used by anything.

yacovm commented on orderer/consensus/smartbft/chain.go:
```diff
@@ -0,0 +1,604 @@
+/*
```
this code has no test. Let's open an issue to write a unit test for this file.

Param-S replied:
https://github.com/hyperledger/fabric/issues/4008

yacovm commented on orderer/consensus/smartbft/chain.go:
```diff
@@ -0,0 +1,604 @@
+// HandleRequest handles the request from the sender
+func (c *BFTChain) HandleRequest(sender uint64, req []byte) {
+	c.Logger.Debugf("HandleRequest from %d", sender)
+	c.consensus.SubmitRequest(req)
```
we need to verify the request. Otherwise a byzantine node may send us a bad  request and we will enqueue it in the block and the followers will detect it, and will start a view change.

yacovm commented on orderer/consensus/smartbft/chain.go:
```diff
@@ -0,0 +1,604 @@
+// discarding the message if the reconfiguration is no longer valid.
+// The consenter may return an error, indicating the message was not accepted
+func (c *BFTChain) Configure(config *cb.Envelope, configSeq uint64) error {
+	// TODO: check configuration update validity
```
I don't think we need this TODO. We are doing the right thing here.

yacovm commented on orderer/consensus/smartbft/chain.go:
```diff
@@ -0,0 +1,604 @@
+		}
+
+		sigs = append(sigs, &cb.MetadataSignature{
+			//	AuxiliaryInput: sig.AuxiliaryInput,
```
please no commented out code.

yacovm commented on orderer/consensus/smartbft/chain.go:
```diff
@@ -0,0 +1,604 @@
+// Errored returns a channel which will close when an error has occurred.
+// This is especially useful for the Deliver client, who must terminate waiting
+// clients when the consenter is not up to date.
+func (c *BFTChain) Errored() <-chan struct{} {
```
let's open an issue to implement this.

yacovm commented on orderer/consensus/smartbft/consenter.go:
```diff
@@ -0,0 +1,271 @@
+		return err == nil
+	})
+
+	consenter.Comm = &cluster.Comm{
```
why are we initializing this with `Comm`?

Aren't we supposed to use the new cluster service?

Param-S replied:
Updated to use new commservice

yacovm commented on orderer/consensus/smartbft/consenter.go:
```diff
@@ -0,0 +1,271 @@
+	return bl.Bytes, nil
+}
+
+func (c *Consenter) detectSelfID(consenters []*cb.Consenter) (uint32, error) {
```
this is not how we should detect our self ID.

Param-S replied:
Updated to use the ecert to detect the node id.

Param-S committed:
addressed first set of review comments

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
Updated detectSelfID to use consentermapping ecert

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

Param-S committed:
BFTchain uses new communication service

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

yacovm committed:
Configure ingress at reconfig and include intermediate TLS CA certs

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Sanitize e-certs in communication mapping and when looking up ids

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Vendor latest consensus library

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Fix iterated removal integration test

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Fix Raft integration tests

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Verify requests received from followers

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Remove redundant print

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Follower detects rogue genesis block, uses BFT signature verification

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Sanitize certs in UT

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Fix UT

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Fix integration test

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Fix UTs, fix some ITs

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
add err check

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm commented on orderer/common/cluster/util.go:
```diff
@@ -348,42 +335,9 @@ func VerifyBlockHash(indexInBuffer int, blockBuff []*common.Block) error {
-		return err
-	}
-	return verifier.VerifyBlockSignature(signatureSet, config)
+func VerifyBlockSignature(block *common.Block, verifier protoutil.BlockVerifierFunc, config *common.ConfigEnvelope) error {
```
we don't use the `config` here... this is a problem.

yacovm committed:
Rebuild verifier when encountering config block

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Remove print, elaborate comment

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm committed:
Resolve module conflicts and rebase on top of main

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

tock-ibm commented on integration/ports.go:
```diff
@@ -41,6 +41,7 @@ const (
 	PrivateDataPurgeBasePort
 	RaftBasePort
 	SBEBasePort
+	SmartBFTBasePort
```
Should the SmartBFTBasePort different than the RaftBasePort? In general, that could be renamed the "ConsensusBasePort"?

tock-ibm commented on integration/channelparticipation/channel_participation.go:
```diff
@@ -42,11 +42,12 @@ func Join(n *nwo.Network, o *nwo.Orderer, channel string, block *common.Block, e
-	err = json.Unmarshal(body, c)
-	Expect(err).NotTo(HaveOccurred())
-	Expect(*c).To(Equal(expectedChannelInfo))
+
+	doBody(client, req)
+
+	Eventually(func() ChannelInfo {
+		return ListOne(n, o, channel)
+	}, n.EventuallyTimeout).Should(Equal(expectedChannelInfo))
```
Why is this change here? 
It is unrelated to BFT, and I think it is wrong. We check the returned body for the correct response. ListOne is done always from the outside later in the tests.,

tock-ibm commented on integration/nwo/configblock.go:
```diff
@@ -12,6 +12,9 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/hyperledger/fabric/common/channelconfig"
+	"github.com/hyperledger/fabric/common/policies"
+
```
goimport

tock-ibm commented on integration/raft/cft_test.go:
```diff
@@ -310,9 +310,9 @@ var _ = Describe("EndToEnd Crash Fault Tolerance", func() {
 				URL:               "/participation/v1/channels/testchannel",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "consenter",
-				Height:            0,
+				Height:            2,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/channel_participation_test.go:
```diff
@@ -140,9 +140,9 @@ var _ = Describe("ChannelParticipation", func() {
 				URL:               "/participation/v1/channels/participation-trophy",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "follower",
-				Height:            0,
+				Height:            3,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/channel_participation_test.go:
```diff
@@ -259,9 +259,9 @@ var _ = Describe("ChannelParticipation", func() {
 				URL:               "/participation/v1/channels/participation-trophy",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "follower",
-				Height:            0,
+				Height:            7,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/channel_participation_test.go:
```diff
@@ -397,9 +397,9 @@ var _ = Describe("ChannelParticipation", func() {
 				URL:               "/participation/v1/channels/participation-trophy",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "consenter",
-				Height:            0,
+				Height:            5,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -316,9 +316,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 				URL:               "/participation/v1/channels/testchannel",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "consenter",
-				Height:            0,
+				Height:            2,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -395,9 +395,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 				URL:               "/participation/v1/channels/testchannel",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "follower",
-				Height:            0,
+				Height:            2,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -839,9 +839,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 				URL:               "/participation/v1/channels/testchannel",
-				Status:            "onboarding",
+				Status:            "active",
 				ConsensusRelation: "consenter",
-				Height:            0,
+				Height:            9,
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -876,9 +876,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 				Name:              "testchannel2",
 				URL:               "/participation/v1/channels/testchannel2",
-				Status:            "onboarding",
+				Status:            "active",
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -1006,9 +1006,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 				Name:              "mychannel",
 				URL:               "/participation/v1/channels/mychannel",
-				Status:            "onboarding",
+				Status:            "active",
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/migration_test.go:
```diff
@@ -424,9 +424,9 @@ var _ = Describe("ConsensusTypeMigration", func() {
 				Name:              channel2,
 				URL:               fmt.Sprintf("/participation/v1/channels/%s", channel2),
-				Status:            "onboarding",
+				Status:            "active",
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -1563,9 +1563,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 					Name:              "testchannel",
 					URL:               "/participation/v1/channels/testchannel",
-					Status:            "onboarding",
+					Status:            "active",
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -1345,9 +1345,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 					Name:              "testchannel",
 					URL:               "/participation/v1/channels/testchannel",
-					Status:            "onboarding",
+					Status:            "active",
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -1033,9 +1033,9 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 				Name:              "mychannel",
 				URL:               "/participation/v1/channels/mychannel",
-				Status:            "onboarding",
+				Status:            "active",
```
This is wrong, because `Join` was changed incorrectly.

tock-ibm commented on integration/nwo/fabricconfig/configtx.go:
```diff
@@ -62,15 +62,24 @@ type Application struct {
 	ExtraProperties map[string]interface{} `yaml:",inline,omitempty"`
 }
 
+type ConsenterMapping struct {
+	ID       uint64
+	Host     string
+	Port     int
+	MSPID    string
+	Identity string
+}
```
Is Raft ignoring the ConsenterMapping? How does it treat this new construct?

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -0,0 +1,1443 @@
+			channel := "testchannel1"
+
+			network = nwo.New(networkConfig, testDir, client, StartPort(), components)
+			network.Consortiums = nil
+			network.Consensus.ChannelParticipationEnabled = true
+			network.Consensus.BootstrapMethod = "none"
```
This should be part of `nwo.MultiNodeSmartBFT()`

tock-ibm commented on integration/nwo/standard_networks.go:
```diff
@@ -270,3 +270,47 @@ func MultiNodeEtcdRaftNoSysChan() *Config {
 	return config
 }
+
+func BasicSmartBFT() *Config {
+	config := BasicConfig()
+	config.Consensus.Type = "BFT"
+	config.Profiles = []*Profile{{
+		Name:     "SampleDevModeSmartBFT",
+		Orderers: []string{"orderer"},
+	}, {
+		Name:          "TwoOrgsChannel",
+		Consortium:    "SampleConsortium",
+		Organizations: []string{"Org1", "Org2"},
+	}}
+	return config
+}
```
disable the system channel stuff here, enable channel participation, etc

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -0,0 +1,1443 @@
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/hyperledger/fabric-config/configtx"
+	"github.com/hyperledger/fabric-config/configtx/orderer"
```
go import

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -0,0 +1,1443 @@
+
+			for _, o := range network.Orderers {
+				By("joining " + o.Name + " to channel as a consenter")
+				channelparticipation.Join(network, o, "testchannel1", genesisBlock, expectedChannelInfoPT)
```
Join was changed incorrectly. Every usage of Join in the new tests should be revisited.

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -0,0 +1,1443 @@
+				Expect(channelInfo).To(Equal(expectedChannelInfoPT))
+			}
+
+			By("Waiting for followers to see the leader")
+			Eventually(ordererRunners[1].Err(), network.EventuallyTimeout, time.Second).Should(gbytes.Say("Message from 1"))
+			Eventually(ordererRunners[2].Err(), network.EventuallyTimeout, time.Second).Should(gbytes.Say("Message from 1"))
+			Eventually(ordererRunners[3].Err(), network.EventuallyTimeout, time.Second).Should(gbytes.Say("Message from 1"))
```
how does that prove the existence of a leader?
who guarantees the leader is orderer 0?
see how it is done in Raft IT

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -0,0 +1,1443 @@
+			ordererProcesses[0].Signal(syscall.SIGTERM)
+			Eventually(ordererProcesses[0].Wait(), network.EventuallyTimeout).Should(Receive())
+
+			By("Waiting for view change to occur")
+			Eventually(ordererRunners[1].Err(), network.EventuallyTimeout*2, time.Second).Should(gbytes.Say("Changing to leader role, current view: 1, current leader: 2 channel=testchannel1"))
+			Eventually(ordererRunners[2].Err(), network.EventuallyTimeout*2, time.Second).Should(gbytes.Say("Changing to follower role, current view: 1, current leader: 2 channel=testchannel1"))
+			Eventually(ordererRunners[3].Err(), network.EventuallyTimeout*2, time.Second).Should(gbytes.Say("Changing to follower role, current view: 1, current leader: 2 channel=testchannel1"))
+			Eventually(ordererRunners[4].Err(), network.EventuallyTimeout*2, time.Second).Should(gbytes.Say("Changing to follower role, current view: 1, current leader: 2 channel=testchannel1"))
```
again, how can you be sure who the leader will be?

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -0,0 +1,1443 @@
+		URL:               "/participation/v1/channels/" + channel,
+		Status:            "active",
+		ConsensusRelation: "consenter",
+		Height:            1,
```
Join was changed incorrectly. Revisit

tock-ibm reviewed (commented):
This is just a quick review of the integration tests.

tock-ibm commented on orderer/consensus/smartbft/util.go:
```diff
@@ -0,0 +1,470 @@
+	}
+
+	for _, consenter := range oc.Consenters() {
+		fmt.Println(base64.StdEncoding.EncodeToString(consenter.ServerTlsCert))
```
leftover from a debug session?
